net/redisclient.go

```diff
@@ -87,6 +100,91 @@ const (
 	defaultConnMetricsInterval = 60 * time.Second
 )
 
+// https://arxiv.org/pdf/1406.2294.pdf
+type jumpHash struct {
+	shards []string
+}
+
+func NewJumpHash(shards []string) redis.ConsistentHash {
+	return &jumpHash{
+		shards: shards,
+	}
+}
+
+func (j *jumpHash) Get(k string) string {
+	hash := fnv.New64()
+	_, err := hash.Write([]byte(k))
+	if err != nil {
+		log.Fatalf("Failed to write %s to hash: %v", k, err)
+	}
+
+	key := hash.Sum64()
+	h := jump.Hash(key, len(j.shards))
+	return j.shards[int(h)]
+}
+
+// Multi-probe consistent hashing - mpchash
+// https://arxiv.org/pdf/1505.00062.pdf
+type multiprobe struct {
+	hash   *mpchash.Multi
+	shards []string
+}
+
+func NewMultiprobe(shards []string) redis.ConsistentHash {
+	return &multiprobe{
+		// 2 seeds and k=21 got from library
+		hash:   mpchash.New(shards, siphash64seed, [2]uint64{1, 2}, 21),
+		shards: shards,
+	}
+}
+
+func (mc *multiprobe) Get(k string) string {
+	return mc.hash.Hash(k)
+}
+func siphash64seed(b []byte, s uint64) uint64 { return siphash.Hash(s, 0, b) }
+
+// rendezvous copied from github.com/go-redis/redis/v8@v8.3.3/ring.go
+type rendezvousWrapper struct {
+	*rendezvous.Rendezvous
+}
+
+func (w rendezvousWrapper) Get(key string) string {
+	return w.Lookup(key)
+}
+
+func NewRendezvous(shards []string) redis.ConsistentHash {
+	return rendezvousWrapper{rendezvous.New(shards, xxhash.Sum64String)}
+}
+
+// rendezvous vnodes
+type rendezvousVnodes struct {
+	*rendezvous.Rendezvous
+	table map[string]string
+}
+
+func (w rendezvousVnodes) Get(key string) string {
+	k := w.Lookup(key)
+	v, ok := w.table[k]
+	if !ok {
+		log.Printf("not found: %s in table for input: %s, so return %s", k, key, v)
+	}
+	return v
+}
+
+func NewRendezvousVnodes(shards []string) redis.ConsistentHash {
+	N := 100
+	vshards := make([]string, N*len(shards))
+	table := make(map[string]string)
+	for i := 0; i < N; i++ {
+		for j, shard := range shards {
+			vshard := fmt.Sprintf("%s%d", shard, i) // suffix
```

Comment:
vshard doesn't depend on j, let's move it out of the inner loop for clarity

Reply:
but depends on shard. Maybe swapping the loop would read better?..
```go
for _, shard := range shards {
  for i := 0; i < N; i++ {
    vshard := fmt.Sprintf("%s%d", shard, i)
  }
}

Reply:
Nevermind

Reply:
but the index in vshards needs j

Reply:
true :)